### PR TITLE
Add functions Add and Del for ObjectMode

### PR DESCRIPTION
**Pull request by JinnyYi**
Add functions `Add` and `Del` for `ObjectMode`.

### CODE REVIEW

File: types/object.go [hunk 1]
```diff
@@ -55,6 +55,12 @@ func (o ObjectMode) String() string {
 	return strings.Join(s, "|")
 }
 
+func (o ObjectMode) Add(mode ObjectMode) {
```
Comment (Xuanwo) [inline types/object.go:58]: Please add comments for this function, like:

```go
// Add support add ObjectMode into current one.
```

File: types/object.go [hunk 2]
```diff
@@ -55,6 +55,12 @@ func (o ObjectMode) String() string {
...
+	o |= mode
+}
+func (o ObjectMode) Del(mode ObjectMode) {
```
Comment (Xuanwo) [inline types/object.go:61]: Please add unit tests and example tests in `object_test.go`

```go
func TestObjectMode_Add(t *testing.T) {
    cases := ...
    for _, v := range cases {
          t.Run(xxxx)
    }
}

func ExampleObjectMode_Add() {
    var o ObjectMode
    o.Add(ModeDir)
    o.Add(ModeRead | ModeAppend)
}
```

File: types/object_test.go [hunk 3]
```diff
@@ -39,3 +39,53 @@ func TestObjectMode_IsDir(t *testing.T) {
...
 }
+
+func TestObjectMode_Add(t *testing.T) {
+	cases := []struct {
+		name   string
+		input  ObjectMode
+		args   ObjectMode
+		expect ObjectMode
+	}{
+		{"simple case", ModeRead, ModeAppend, ModeRead | ModeAppend},
```
Comment (Xuanwo) [inline types/object_test.go:50]: Please add more case:

- What if `Read` + `Read`?
- What if `Read` + `Append|Read`

File: types/object_test.go [hunk 4]
```diff
@@ -39,3 +39,53 @@ func TestObjectMode_IsDir(t *testing.T) {
 		})
 	}
 }
+
+func TestObjectMode_Add(t *testing.T) {
+	cases := []struct {
+		name   string
+		input  ObjectMode
```
Comment (Xuanwo) [inline types/object_test.go:46]: maybe rename `input` to `base` and  `args` to `input` for better understanding?

File: types/object_test.go [hunk 5]
```diff
@@ -39,3 +39,53 @@ func TestObjectMode_IsDir(t *testing.T) {
...
+			v.input.Add(v.args)
+			assert.Equal(t, v.expect, v.input)
+		})
+	}
+}
+
+func TestObjectMode_Del(t *testing.T) {
+	cases := []struct {
+		name   string
+		input  ObjectMode
```
Comment (Xuanwo) [inline types/object_test.go:65]: ditto

File: types/object_test.go [hunk 6]
```diff
@@ -39,3 +39,53 @@ func TestObjectMode_IsDir(t *testing.T) {
...
+		args   ObjectMode
+		expect ObjectMode
+	}{
+		{"simple case", ModeRead | ModeAppend, ModeAppend, ModeRead},
```
Comment (Xuanwo) [inline types/object_test.go:69]: Please add more case:

- What if `Read` - `Read`?
- What if `Read | Dir` - `Read | Append`?

File: types/object.go [hunk 7]
```diff
@@ -55,10 +55,17 @@ func (o ObjectMode) String() string {
...
 }
 
+// Add support add ObjectMode into current one.
 func (o ObjectMode) Add(mode ObjectMode) {
 	o |= mode
 }
+
+// Del support delete ObjectMode from current one.
 func (o ObjectMode) Del(mode ObjectMode) {
+	if o&mode == 0 {
```
Comment (Xuanwo) [inline types/object.go:65]: Is the check required?

File: types/object_test.go [hunk 8]
```diff
@@ -39,3 +39,53 @@ func TestObjectMode_IsDir(t *testing.T) {
...
+		{"complex case", ModeDir, ModeRead, ModeDir},
+	}
+
+	for _, v := range cases {
+
```
Comment (Xuanwo) [inline types/object_test.go:74]: Please remove this empty line.

File: types/object.go (hunk 2 above)
Comment (Xuanwo) [reply types/object.go:64]: In order to check value at local, we may need to change `(o ObjectMode) Del` to `(o *ObjectMode) Del`

Comment (Xuanwo) [reply types/object.go:64]: We may need to change `o ObjectMode` into `o *ObjectMode`

File: types/object.go (hunk 7 above)
Comment (JinnyYi) [reply types/object.go]: No, it's unneccesary.

File: types/object_test.go [hunk 9]
```diff
@@ -39,3 +39,57 @@ func TestObjectMode_IsDir(t *testing.T) {
...
+func TestObjectMode_Add(t *testing.T) {
+	cases := []struct {
+		name   string
+		base   ObjectMode
+		input  ObjectMode
+		expect ObjectMode
+	}{
+		{"simple case", ModeRead, ModeAppend, ModeRead | ModeAppend},
+		{"complex case", ModeDir, ModeLink | ModeRead, ModeDir | ModeLink | ModeRead},
+		{"complex case", ModeRead, ModeRead, ModeRead},
```
Comment (Xuanwo) [inline types/object_test.go:52]: Please give a meaningful name, like: "add new mode", "add exist mode"

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**